src/Common/Coverage.h

```diff
@@ -0,0 +1,272 @@
+
+    const std::string logger_base_name {"Coverage"};
+
+    const size_t hardware_concurrency { std::thread::hardware_concurrency() };
+    const Poco::Logger * base_log {nullptr}; // Initialized when server initializes Poco internal structures.
+
+    const SymbolIndexInstance symbol_index;
+    const Dwarf dwarf;
+
+    const std::string report_path { docker_report_path };
+
+    // CH client is usually located inside main CH binary, but we don't need to instrument client code.
+    // This variable is set on client initialization so we can ignore coverage for it.
+    bool is_client {false};
+
+    size_t functions_count {0}; // Initialized after processing PC table.
+    size_t addrs_count {0};
+    size_t edges_count {0};
+
+    TaskQueue tasks_queue;
+
+    using SourceRelativePath = std::string; // from ClickHouse src/ directory
+
+    struct SourceFileInfo
+    {
+        SourceRelativePath relative_path;
+        std::vector<EdgeIndex> instrumented_functions;
+        std::vector<Line> instrumented_lines;
+
+        explicit SourceFileInfo(std::string path_)
+            : relative_path(std::move(path_)), instrumented_functions(), instrumented_lines() {}
+    };
+
+    using SourceFileIndex = size_t;
+
+    struct EdgeInfo
+    {
+        Line line;
+        SourceFileIndex index;
+        std::string_view name;
+
+        constexpr bool isFunctionEntry() const { return !name.empty(); }
+    };
+
+    struct TestData
+    {
+        struct SourceData
+        {
+            std::vector<EdgeIndex> functions_hit;
+            std::vector<Line> lines_hit;
+        };
+
+        std::string name; // If is empty, there is no data for test.
+        size_t test_index;
+        const Poco::Logger * log;
+        std::vector<SourceData> data; // [i] ~ info for source_files_cache[i].
+    };
+
+    /// [i] ~ is edge with index i hit or not.
+    /// Can't use vector<bool> as it's not contiguous.
+    using EdgesHit = std::vector<char>;
+
+    std::vector<Addr> edges_to_addrs;
+    std::vector<bool> edge_is_func_entry;
+
+    std::vector<SourceFileInfo> source_files_cache;
+    std::unordered_map<SourceRelativePath, SourceFileIndex> source_rel_path_to_index;
+
+    std::vector<EdgeInfo> edges_cache;
+
+    std::vector<TestData> tests; /// Data accumulated for all tests.
+    EdgesHit edges_hit; /// Data accumulated for a single test.
+
+    size_t test_index {0}; /// Index for test that will run next.
+
+    class FileWrapper
+    {
+        FILE * handle {nullptr};
+    public:
+        inline void set(const std::string& pathname, const char * mode) { handle = fopen(pathname.data(), mode); }
+        inline FILE * file() { return handle; }
+        inline void close() { fclose(handle); handle = nullptr; }
+        inline void write(const fmt::memory_buffer& mb) { fwrite(mb.data(), sizeof(char), mb.size(), handle); }
+    };
+
+    FileWrapper report_file;
```

Comment:
Why we use custom FileWrapper implementation, cant we use WriteBufferFromFile ?

Reply:
No, otherwise we would need to exclude this file from instrumentation.